lib/train/transports/cisco_ios.rb

```diff
@@ -0,0 +1,141 @@
+# encoding: utf-8
+
+require 'train/plugins'
+require 'train/transports/ssh'
+
+module Train::Transports
+  class BadEnablePassword < Train::TransportError; end
+
+  class CiscoIOS < SSH
+    name 'cisco_ios'
+
+    option :host, required: true
+    option :user, required: true
+    option :port, default: 22, required: true
+
+    option :password, required: true
+
+    # Used to elevate to enable mode (similar to `sudo su` in Linux)
+    option :enable_password
+
+    def connection
+      @connection ||= Connection.new(validate_options(@options).options)
+    end
+
+    class Connection < BaseConnection
+      def initialize(options)
+        super(options)
+
+        @session = nil
+        @buf = nil
+
+        # Delete options to avoid passing them in to `Net::SSH.start` later
+        @host = @options.delete(:host)
+        @user = @options.delete(:user)
+        @port = @options.delete(:port)
+        @enable_password = @options.delete(:enable_password)
+
+        @prompt = /^\S+[>#]\r\n.*$/
+      end
+
+      def uri
+        "ssh://#{@user}@#{@host}:#{@port}"
+      end
+
+      private
+
+      def establish_connection
+        logger.debug("[SSH] opening connection to #{self}")
+
+        Net::SSH.start(
+          @host,
+          @user,
+          @options.reject { |_key, value| value.nil? },
+        )
+      end
+
+      def session
+        return @session unless @session.nil?
+
+        @session = open_channel(establish_connection)
+
+        # Escalate privilege to enable mode if password is given
+        if @enable_password
+          run_command_via_connection("enable\r\n#{@enable_password}")
+        end
+
+        # Prevent `--MORE--` by removing terminal length limit
+        run_command_via_connection('terminal length 0')
+
+        @session
+      end
+
+      def run_command_via_connection(cmd)
+        # Ensure buffer is empty before sending data
+        @buf = ''
+
+        logger.debug("[SSH] Running `#{cmd}` on #{self}")
+        session.send_data(cmd + "\r\n")
+
+        logger.debug('[SSH] waiting for prompt')
+        until @buf =~ @prompt
+          raise BadEnablePassword if @buf =~ /Bad secrets/
+          session.connection.process(0)
+        end
+
+        # Save the buffer and clear it for the next command
+        output = @buf.dup
+        @buf = ''
+
+        result = format_output(output, cmd)
+        CommandResult.new(*format_result(result))
+      end
+
+      ERROR_MATCHERS = [
+        'Bad IP address',
+        'Incomplete command',
+        'Invalid input detected',
+        'Unrecognized host',
+      ].freeze
+
+      def format_result(result)
+        stderr_with_exit_1 = ['', result, 1]
+        stdout_with_exit_0 = [result, '', 0]
```

Comment:
I put my finger on why these bother me. One of them is always a spurious assignment. At this point I'd definitely either extract them as methods (`format_err(result)`, `format_out(result)`) or inline the arrays, or inline `CommandResult.new()`.

```
def format_result(result)
  if ERROR_MATCHERS.none? { |e| result.include?(e) }
    CommandResult.new(result, '', 0)
  else
    CommandResult.new('', result, 1)
  end
end
```
I do wish there were `CommandResult` subclasses that revealed intent though. `ErrorResult.new(result)` which automatically assigns `result` to stderr and returns 1, etc.

Reply:
I'm glad you put a finger on it. I took your code suggestion and ran with it. Let me know if it's any better.